src/main/java/net/neoforged/serverstarterjar/Main.java

```diff
@@ -166,51 +170,51 @@ public static void mainLaunch(String[] starterArgs) throws Throwable {
             var cp = readClasspathAttribute(jarFile);
             cp.add(0, jarFile.toPath());
 
-            // Update the java.class.path sys prop with the jar and its Class-Path
-            var cpProperty = new StringBuilder(System.getProperty("java.class.path"));
-
-            var systemCl = ClassLoader.getSystemClassLoader();
-            for (Path path : cp) {
-                var absolute = path.toAbsolutePath().toString();
-                cpProperty.append(File.pathSeparatorChar).append(absolute);
-                appendClassPath.invoke(systemCl, absolute);
-            }
-
-            System.setProperty("java.class.path", cpProperty.toString());
+            addToClassPath(cp);
         }
 
-        // Otherwise, go back to trying to find the module path
+        // Otherwise, go back to trying to find the module or classpath
         else {
-            final var pathArg = findValue(args, "-p");
-            if (pathArg == null) {
-                System.err.println("Could not find module path (specified by -p)");
-                System.exit(1);
-                return;
-            }
+            loadJavaAgents(args);
```

Comment:
I would split half of the logic in `loadJavaAgents` so that we can load the premain of the `-jar` too for full support.